app/models/job_proxy_dispatcher.rb

```diff
@@ -144,21 +194,37 @@ def assign_proxy_to_job(proxy, job)
     end
   end
 
-  def pending_jobs
+  def pending_jobs(target_class = VmOrTemplate)
+    class_name = target_class.base_class.name
     @zone = MiqServer.my_zone
     Job.order(:id)
-      .where(:state           => "waiting_to_start")
-      .where(:dispatch_status => "pending")
-      .where("zone is null or zone = ?", @zone)
-      .where("sync_key is NULL or
-        sync_key not in (
+       .where(:state           => "waiting_to_start")
+       .where(:dispatch_status => "pending")
+       .where(:target_class    => class_name)
+       .where("zone is null or zone = ?", @zone)
+       .where("sync_key is NULL or
+         sync_key not in (
            select sync_key from jobs where
              dispatch_status = 'active' and
              state != 'finished' and
              (zone is null or zone = ?) and
              sync_key is not NULL)", @zone)
   end
 
+  def pending_container_jobs
+    jobs = pending_jobs(ContainerImage)
+    involved_container_images = jobs.collect(&:target_id).uniq
+    container_images_to_ems, jobs_by_ems = {}, {}
+    ContainerImage.where(:id => involved_container_images).each do |container_image|
+      container_images_to_ems[container_image.id] = container_image.ems_id
+    end
+    jobs.each do |job|
+      jobs_by_ems[container_images_to_ems[job.target_id]] ||= []
+      jobs_by_ems[container_images_to_ems[job.target_id]] << job
+    end
+    [jobs_by_ems, container_images_to_ems]
```

Comment:
I wouldn't be offended if you thought it was better to return objects over hashes here and change the callers to call methods on those objects instead of everything knowing the structure of these hashes.  Since you're the only consumer of this method, feel free to NOT use existing hash structures.  Or clean this up in a followup PR.

Hashes are convenient but leak internal structure, making it hard to change things.  Hence, the existing job proxy dispatcher code base.